### PR TITLE
Fix GetVersion

### PR DESCRIPTION
**Pull request by akawashiro**
This is the second PR which I mentioned in https://github.com/shinh/sold/pull/34#discussion_r505131071.

GetVersion have returned a pair of (filename, version) when it finds the corresponding entry in Elf_Verneed. This PR fixes GetVersion so that it returns a pairt of (soname, version) always.

### CODE REVIEW

File: elf_binary.cc [hunk 1]
```diff
@@ -181,7 +181,14 @@ std::pair<std::string, std::string> ELFBinary::GetVersion(int index) {
                     if (vna->vna_other == versym_[index]) {
                         LOG(INFO) << "Find Elf_Vernaux corresponds to " << versym_[index] << SOLD_LOG_KEY(strtab_ + vn->vn_file)
                                   << SOLD_LOG_KEY(strtab_ + vna->vna_name);
-                        return std::make_pair(std::string(strtab_ + vn->vn_file), std::string(strtab_ + vna->vna_name));
+
+                        std::string filename = std::string(strtab_ + vn->vn_file);
```
Comment (shinh) [inline elf_binary.cc:185]: Please do not call `map::find` twice. `auto found = filename_to_soname.find(filename);`?

**Review by shinh — COMMENTED** [review]
Could you explain why this PR is necessary, instead of what you have done? It is hard for me to determine if this PR is changing to the right direction or not. I need to carefully read glibc by myself to understand this is a legitimate change, which takes not small time. I think you have already studied a lot. Sharing what you learned will help me to understand PRs. It'll be a great reference in future, too.

Thanks!!

**Comment by akawashiro** [comment]
Sorry for my insufficient explanation.

GetVersion has returned a pair of (filename, version) when it finds the corresponding entry in Elf_Verneed although it has returns (soname, version) when it finds in Elf_Verdef. Because of this behavior, ELFBinary::syms_ can contain an entry whose soname field is filename, not soname. As a result, SymtabBuilder::Resolve doesn't work when the filename is different from the soname.

There are two ways to fix this bug. One is enforcing GetVersion to return (filename, version), another is enforcing GetVersion to return (soname, version).  The difference between the two ways is how to handle two shared objects with different filenames but with the same soname.

With the first one, symbols are resolved considering the difference of filenames. However, with the second one,  symbols are resolved without considering the difference.

I think the second one is preferable.  Because [check_match  in dl-lookup.c](https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/elf/dl-lookup.c#L121) doesn't using the filename field to match the symbol.

File: elf_binary.cc (hunk 1 above)
Comment (akawashiro) [reply elf_binary.cc:185]: I see. I'll fix it.

File: elf_binary.cc [hunk 2]
```diff
@@ -181,7 +181,15 @@ std::pair<std::string, std::string> ELFBinary::GetVersion(int index) {
...
+                        auto found = filename_to_soname.find(filename);
+                        if (found != filename_to_soname.end()) {
+                            return std::make_pair(found->second, std::string(strtab_ + vna->vna_name));
+                        } else {
+                            LOG(WARNING) << "There is no entry for " << filename << " in filename_to_soname. Use filename as soname. ";
```
Comment (shinh) [inline elf_binary.cc:190]: Can this code path actually happen? Shouldn't we always have an entry for all filenames?

File: sold.cc [hunk 3]
```diff
@@ -821,6 +822,17 @@ class Sold {
         }
     }
 
+    void ConstructFilenameToSoname() {
+        for (const auto& l : link_binaries_) {
+            if (l->name() != "" && l->soname() != "") {
+                filename_to_soname_[l->name()] = l->soname();
+                LOG(INFO) << SOLD_LOG_KEY(l->name()) << SOLD_LOG_KEY(l->soname());
+            } else {
+                LOG(WARNING) << "Empty filename or soname: " << SOLD_LOG_KEY(l->name()) << SOLD_LOG_KEY(l->soname());
```
Comment (shinh) [inline sold.cc:831]: Is the a valid code path? Shouldn't we always have both info?

File: sold.cc [hunk 4]
```diff
@@ -821,6 +822,17 @@ class Sold {
         }
     }
 
+    void ConstructFilenameToSoname() {
+        for (const auto& l : link_binaries_) {
```
Comment (shinh) [inline sold.cc:826]: I think it's better to check if `soname` in `link_binaries_` are unique.

**Review by shinh — COMMENTED** [review]
Thanks for your explanation! Now I feel more comfortable with this PR.

> how to handle two shared objects with different filenames but with the same soname.

I thought having two shared objects with the same soname is invalid. Do you actually have a case where this is the case?  It seems the linker does not like it at least? 

```
#!/bin/bash

set -ex

cd /tmp

cat <<EOF > func1.c
int func1() { return 1; }
EOF
gcc func1.c -shared -Wl,-soname=libfunc -o libfunc1.so

cat <<EOF > func2.c
int func2() { return 2; }
EOF
gcc func2.c -shared -Wl,-soname=libfunc -o libfunc2.so

cat <<EOF > main.c
int func1();
int func2();
int main() {
  return func1() + func2();
}
EOF
gcc main.c libfunc1.so libfunc2.so
```

File: elf_binary.cc (hunk 2 above)
Comment (akawashiro) [reply elf_binary.cc]: When a shared object (e.g., libc.so) is supposed not to link by ShouldLink, 
I didn't push its name to  Sold::filename_to_soname_. To overcome this bug, we needed this code path.

But I fixed this bug with https://github.com/shinh/sold/pull/37/commits/527fbd451730916b00de9b6b96ff5c17df5157ca, So I deleted the code path in https://github.com/shinh/sold/pull/37/commits/e99016fa640c84036bee0f8b4904805c1008fc6a.

(Although I said "some shared objects don't have a soname" at today's meeting, it's only my mistake.)

File: sold.cc (hunk 3 above)
Comment (akawashiro) [reply sold.cc]: Although shared objects have non-empty soname and filename always, executable files, such as a.out, may not have DT_SONAME fields. So, I decided to allow empty sonames only for Sold::main_binary_. See https://github.com/shinh/sold/pull/37/commits/527fbd451730916b00de9b6b96ff5c17df5157ca.

**Comment by akawashiro** [comment]
In the later PR, I will make sold not to load shared objects with the same soname. (Now, I am reading glibc to find the correct order of searching shared objects. BFS, DFS, or something else...)

File: sold.cc (hunk 4 above)
Comment (akawashiro) [reply sold.cc]: I do it in the future PR when I change the order of loading.

**Review by shinh — APPROVED** [review]
LGTM!!

> I am reading glibc to find the correct order of searching shared objects. BFS, DFS, or something else

I think it's in BFS order: http://www.skyfree.org/linux/references/ELF_Format.pdf I'm not 100% sure as this is about the order of symbol resolution, though.

> When resolving symbolic references, the dynamic linker examines the symbol tables with a breadth-first search. That is, it first looks at the symbol table of the executable program itself, then at the symbol tables of the DT_NEEDED entries (in order), then at the second level DT_NEEDED entries, and so on.

**Comment by akawashiro** [comment]
Thank you.

**Pull request merged**